README.md
[![CircleCI](https://circleci.com/gh/ConsenSys/truffle-security.svg?style=svg)](https://circleci.com/gh/ConsenSys/truffle-security)
[![Coverage Status](https://coveralls.io/repos/github/ConsenSys/truffle-security/badge.svg?branch=master)](https://coveralls.io/github/ConsenSys/truffle-security?branch=master)
[![ESDoc](https://doc.esdoc.org/github.com/ConsenSys/truffle-security/badge.svg)](https://doc.esdoc.org/github.com/ConsenSys/truffle-security)

# MythX Security Analysis Plugin for Truffle Framework

This plugin adds automated smart contract security analysis to the [Truffle framework](https://truffleframework.com/docs/truffle/overview). It is based on [MythX](https://mythx.io), the security analysis API for Ethereum smart contracts. The plugin is compatible with Truffle 5.0 or higher.

## Installing the Plugin

To install the latest stable version from NPM:

```console
$ npm install -g truffle-security
```

### Windows only
On Windows `node-gyp` dependency requires `windows-build-tools` to be installed from an elevated PowerShell or CMD.exe (run as Administrator).
```console
npm install --global --production windows-build-tools
```
For more details refer to [node-gyp installation guide](https://github.com/nodejs/node-gyp#option-1).

## Configuration

Currently, the plugin must be activated on a per-project basis. If `truffle-security` was installed to the Truffle project root, it will try to automatically install itself to `truffle-config.js`. If you installed `truffle-security` globally, add the following to `truffle-config.js` in the root directory of your Truffle project to enable the plugin:

```javascript
module.exports = {
    plugins: [ "truffle-security" ]
};
```

### MythX Account

You can set up a free account on the [MythX website](https://mythx.io) to get full access. Generate your API key in the tools section of the [MythX dashboard](https://dashboard.mythx.io/#/). 

The key can be passed to Truffle either via the `MYTHX_API_KEY` environment variable or the `--apiKey` command line argument. For security reasons it is recommended to always pass the token through an environment variable, e.g. defined in the settings of a Continuous Integration (CI) server or a shell script that can be sourced from.

Set the following enviromment variables to your API key (add this to your `.bashrc` or `.bash_profile` for added convenience):

```bash
export MYTHX_API_KEY='Put your API key here!'
```

And if you're using Windows OS with PowerShell:

```bash
$env:MYTHX_API_KEY='Put your API key here!'
```

### Solc Version

You can specify which version of solc to use in `truffle-config.js` as explained in [truffle's documentation](https://truffleframework.com/docs/truffle/reference/configuration#solc). MythX for Truffle will use the same version of solc that Truffle uses to compile and analyze your contracts.

```
module.exports = {
  plugins: [ "truffle-security" ],
  networks: {
    ... etc ...
  },
  compilers: {
     solc: {
       version: <string>  // ex:  "0.4.20". (Default: Truffle's installed solc)
     }
  }
};
```

## Running Security Analyses

Once the plugin is installed the `truffle run verify` becomes available. You can either analyze a specific file by running `truffle run verify <file-name>`, a contract by running `truffle run verify <file-name>:<contract-name>`, or the entire project with simply `truffle run verify`.

Alternatively you can use `truffle run mythx` instead of `truffle run verify`.

**Your project must compile successfully for the security analysis to work.** Note that the `verify` command invokes `truffle compile` automatically if the build files are not up to date.

Here is the output of `truffle verify` for an [example](https://github.com/ConsenSys/mythx-playground/tree/master/exercise2) from the [DevCon4 MythX Workshop](https://github.com/ConsenSys/mythx-workshop):

```console
$ truffle run verify

/Projects/mythx-playground/exercise2/contracts/Tokensale.sol
   1:0   warning  A floating pragma is set                SWC-103
  16:29  warning  The binary multiplication can overflow  SWC-101
  18:8   warning  The binary addition can overflow        SWC-101

✖ 4 problems (0 errors, 4 warnings)
```

Here is an example of analyzing the same contract passing it in directly and using the `table` report style:

```
$ truffle run verify contracts/Tokensale.sol:Tokensale --style table

/Projects/mythx-playground/exercise2/contracts/Tokensale.sol

║ Line     │ Column   │ Type     │ Message                                                │ Rule ID              ║
╟──────────┼──────────┼──────────┼────────────────────────────────────────────────────────┼──────────────────────╢
║ 1        │ 0        │ warning  │ A floating pragma is set.                              │ SWC-103              ║
║ 16       │ 29       │ warning  │ The binary multiplication can overflow.                │ SWC-101              ║
║ 18       │ 8        │ warning  │ The binary addition can overflow.                      │ SWC-101              ║

╔════════════════════════════════════════════════════════════════════════════════════════════════════════════════╗
║ 0 Errors                                                                                                       ║
╟────────────────────────────────────────────────────────────────────────────────────────────────────────────────╢
║ 4 Warnings                                                                                                     ║
╚════════════════════════════════════════════════════════════════════════════════════════════════════════════════╝

```

## Advanced Options

Run `truffle run verify --help` to show advanced configuration options.

```console
$ truffle run verify --help

Usage: truffle run verify [options] [solidity-file[:contract-name] [solidity-file[:contract-name] ...]]

Runs MythX analyses on given Solidity contracts. If no contracts are
given, all are analyzed.

Options:
  --all      Compile all contracts instead of only the contracts changed since last compile.
  --apiClient { mythxjs | armlet}
            Which api client to use. Default and recommended is mythxjs.
  --mode { quick | standard | deep}
             Perform quick, in-depth (standard) or deep analysis. Default = quick.
  --style { stylish | json | table | tap | unix | markdown | ... },
             Output report in the given es-lint style style.
             See https://eslint.org/docs/user-guide/formatters/ for a full list.
             The markdown format is also included.
  --json | --yaml
             Dump results in unprocessed JSON or YAML format as it comes back from MythX.
             Note: this disables providing any es-lint style reports, and that
             --style=json is processed for eslint, while --json is not.
  --timeout *secs*
             Limit MythX analyses time to *secs* seconds.
             The default is 300 seconds (five minutes).
  --initial-delay *secs*
             Minimum amount of time to wait before attempting a first status poll to MythX.
             The default is 45 seconds.
             See https://github.com/ConsenSys/armlet#improving-polling-response
  --limit *N*
             Have no more than *N* analysis requests pending at a time.
             As results come back, remaining contracts are submitted.
             The default is 4 contracts, the maximum value, but you can
             set this lower.
  --debug    Provide additional debug output. Use --debug=2 for more
             verbose output
  --min-severity { warning | error }
             Ignore SWCs below the designated level
  --swc-blacklist { 101 | 103,111,115 | ... }
             Ignore a specific SWC or list of SWCs.
  --uuid *UUID*
             Print in YAML results from a prior run having *UUID*
             Note: this is still a bit raw and will be improved.
  --version  Show package and MythX version information.
  --progress, --no-progress
             Enable/disable progress bars during analysis. The default is enabled.
  --mythx-logs --no-mythx-logs
             Enable/disable  MythX logs.
  --ci
             Blocking non zero return for CI integrations to throw an error (non-zero exit code).
  --ci-whitelist { 101 | 103,111,115 | ... }
             List of allowed SWCs that will not throw an error (non-zero exit code).
  --apiKey { api key generated from profile dashboard}
             Authenticate with api key instead of login details.
  --color, --no-color
             Enable/disable output coloring. The default is enabled.


```
Configuration options can also be stored as json in `truffle-security.json` at the truffle project root. i.e. : 
```
{
    "style": "table",
    "mode": "quick",
    "min-severity": "warning",
    "swc-blacklist": [103,111]
}
```


NEWS.md
Release 1.7.2
================
Update parser to work with solcjs versions above v0.6.0

Release 1.7.0
================

Can now use MYTHX_API_KEY environment variable for passing api key.

Can now use MYTHX_USERNAME for authentication.

Support for new MythX plans.

Removal of trial mode.


Release 1.6.2
================

Grouping analyses into batches with links to relevant page on dashboard has been added.

Added --no-mythx-logs command to remove mythx logs from Truffle Security output.

Fix output styles.

Updated all truffle packages.

Added --ci and --ci-whitelist for throwing an error based on results.

Can now use `truffle run mythx` as an alternative to `truffle run verify`.

Cancelling during full mode analysis will now return a link to group on the dashboard.

MYTHX_API_URL environment variable is now passed to mythxjs.

Can now pass api key generated from dashboard with --apiKey {apiKey}.


Release 1.6.1
================

Minor bug fixes

Release 1.6.0
================

Default apiClient is now MythXJS to use armlet run `truffle run verify --apiClient armlet`. This will be deprecated in future.

Various bug fixes.

Test updates.

=======
Release 1.5.2
================

Fixed an issue causing issues to be reported at incorrect contracts.

Verify cli arguments better.

Fixed SWC links in html report.

Updated role names to have the same terminology as the api.

Release 1.5.1
================

Increased timeout duration for full mode.

Fixed analysis of contracts with relative imports on Windows.


Release 1.5.0
================

Added a markdown eslint formatter `truffle run verify --style markdown` for markdown reports.

Fixed an error caused by windows filepaths with backslashes.

Fixed a bug preventing analyses with `--mode full`.

Improved error messages when imported contracts fail to compile.

Release 1.4.4
================

Fix an issue causing integration tests to fail.

Release 1.4.3
================

Changed interface from:
```
truffle run verify ContractName
```
To:
```
truffle run verify contracts/contract.sol
```
or
```
truffle run verify contracts/contract.sol:ContractName
```
(ContractName still works, but is deprecated)

Refactored how Truffle build objects are chosen so that the correct ones are always used.

Fixed a bug with sourceList ordering being unreliable, causing line numbers to be reported incorrectly.

Fixed requirement for evmVersion in truffle-compile.

Release 1.4.2
================

Added another missing dependency.

Release 1.4.1
================

Added a missing dependency.

Release 1.4.0
================

Fixed an issue sending bad data to the API.

Fixed a bug compiling solidity files without contracts.

Release 1.3.7
================

Fixed a Windows compatibility issue caused by forward slashes in file paths.

Made minor improvements to error messaging for compilation errors.

Fixed support for running MythX for Truffle with a globally installed npm. This improvement was kindly contributed by Teruhiro Tagomori at [NRI Secure Technologies](https://www.nri-secure.com/security-consulting/blockchain-assessment).

Release 1.3.6
================

We now send the complete AST and sources, including imports to the api, resulting in fewer errors and providing more meaningful output.

Fixed a bug that caused incorrect line numbers to be reported when contracts imported a vulnerable contract.

Fixed an issue that caused the compiler to fail silently or with a confusing error message.

Improved the messaging regarding a missing config file.

Other minor bugfixes.

Release 1.3.5
================

We now prompt users to sign up when they are using the trial account, and provide warnings that not all issues are reported. In addition, we identify whether an account is a Free or Premium account.

We added the `mainSource` field to the data sent to the api, allowing Maru to return better analysis and fewer errors.

Documentation improvements, especially for Windows installation.

Merged in some fixes from the `truffle-compile` repository.

Fixed a regression that caused the compiler to download once for every contract analyzed again.

Fixed an issue that prevented tracebacks from being printed when there are bugs during compilation.

Release 1.3.4
================

Fix a serious bug involving resolve/reject params being swapped.

Release 1.3.3
================

Display the issue warning that the Trial version is being used.

Fixed a bug that prevented analysis of multiple contracts with the same name.

Fixed a bug that downloaded the compiler for every contract analyzed.

Much of this work was kindly contributed by Teruhiro Tagomori at [NRI Secure Technologies](https://www.nri-secure.com/security-consulting/blockchain-assessment).

Release 1.3.2
================

We now automatically install truffle-security to the truffle project it is being installed to.

Support for EthPM packages was added.

Bugs in the `--yaml` output format have been fixed.

Minor logging improvements.

Fixed a bug with a "No issues found" response.

A lot of this work was kindly contributed by Teruhiro Tagomori at [NRI Secure Technologies](https://www.nri-secure.com/security-consulting/blockchain-assessment).

Release 1.3.1
================

We now handle imports from locally-installed NPM packages, such as is how openzepplin is usually installed.

In future releases we start handling:
* EthPM packages
* globally installed NPM packages
* EIP 82

The `--debug` flag now turns off the progress bar. The two, debug information and the progress bar, are incompatible.

MythX log "info" log-level messages are now ignored unless `--debug` is set.

Teruhiro Tagomori at [NRI Secure Technologies](https://www.nri-secure.com/security-consulting/blockchain-assessment) contributed on every aspect of this release.

Release 1.3.0
================

MythX-specific Artifacts
-----------------------------


helpers.js
'use strict';

const armlet = require('armlet');
const armletClass = require('./classes/armlet');
const mythxjsClass = require('./classes/mythx');



const trialEthAddress = '0x0000000000000000000000000000000000000000';
const trialPassword = 'trial';
const defaultAnalyzeRateLimit = 4;
const defaultAPIClient = 'mythxjs';

let client;
/**
 *
 * @param {Object} config - truffle configuration object.
 */
async function analyze(config) {
    config = prepareConfig(config);
    if (config.apiClient === 'armlet') {
        console.log('WARNING: You are using Armlet we will be deprecating Armlet in future versions of truffle-security in favour of MythXJS as it is no longer supported.')
        client = new armletClass( config, 'truffle');
    }
    else {
        client = new mythxjsClass( config, 'truffle');
    }
    const analysisResponse = await client.analyze();
    return analysisResponse;

}

/**
 *
 * Returns a JSON object from a version response. Each attribute/key
 * is a tool name and the value is a version string of the tool.
 *
 * @param {Object} jsonResponse
 * @returns string  A comma-separated string of tool: version
 */
function versionJSON2String(jsonResponse) {
    return Object.keys(jsonResponse).sort().map((key) => `${key}: ${jsonResponse[key]}`).join(', ');
}

/**
 *
 * Handles: truffle run verify --help
 *
 * @returns promise which resolves after help is shown
 */
function printHelpMessage() {
    return new Promise(resolve => {
        const helpMessage = `Usage: truffle run verify [options] [solidity-file[:contract-name] [solidity-file[:contract-name] ...]]

Runs MythX analyses on given Solidity contracts. If no contracts are
given, all are analyzed.

Options:
  --all      Compile all contracts instead of only the contracts changed since last compile.
  --apiClient { mythxjs | armlet}
              Which api client to use. Default and recommended is mythxjs.
  --mode { quick | standard | deep}
             Perform quick, in-depth (standard) or deep analysis. Default = quick.
  --style { stylish | json | table | tap | unix | markdown | ... },
             Output report in the given es-lint style style.
             See https://eslint.org/docs/user-guide/formatters/ for a full list.
             The markdown format is also included.
  --json | --yaml
             Dump results in unprocessed JSON or YAML format as it comes back from MythX.
             Note: this disables providing any es-lint style reports, and that
             --style=json is processed for eslint, while --json is not.
  --timeout *secs*
             Limit MythX analyses time to *secs* seconds.
             The default is 300 seconds (five minutes).
  --initial-delay *secs*
             Minimum amount of time to wait before attempting a first status poll to MythX.
             The default is ${armlet.defaultInitialDelay / 1000} seconds.
             See https://github.com/ConsenSys/armlet#improving-polling-response
  --limit *N*
             Have no more than *N* analysis requests pending at a time.
             As results come back, remaining contracts are submitted.
             The default is ${defaultAnalyzeRateLimit} contracts, the maximum value, but you can
             set this lower.
  --debug    Provide additional debug output. Use --debug=2 for more
             verbose output
             Note: progress is disabled if this is set.
  --min-severity { warning | error }
             Ignore SWCs below the designated level
  --swc-blacklist { 101 | 103,111,115 | ... }
             Ignore a specific SWC or list of SWCs.
  --uuid *UUID*
             Print in YAML results from a prior run having *UUID*
             Note: this is still a bit raw and will be improved.
  --version  Show package and MythX version information.
  --progress, --no-progress
             Enable/disable progress bars during analysis. The default is enabled.
             Note: this is disabled if debug is set.
  --mythx-logs --no-mythx-logs
             Enable/disable  MythX logs.
  --ci
             Blocking non zero return for CI integrations to throw an error (non-zero exit code).
  --ci-whitelist { 101 | 103,111,115 | ... }
             List of allowed SWCs that will not throw an error (non-zero exit code).
  --apiKey { api key generated from profile dashboard}
             Authenticate with api key instead of login details.
  --color, --no-color
             Enable/disable output coloring. The default is enabled.
`;
        // FIXME: decide if this is okay or whether we need
        // to pass in `config` and use `config.logger.log`.
        console.log(helpMessage);
        resolve(null);
    });
}

/**
 *
 * Handles: truffle run verify --version
 * Shows version information for this plugin and each of the MythX components.
 *
 * @returns promise which resolves after MythX version information is shown
 */
async function printVersion() {
    const pjson = require('./package.json');
    // FIXME: decide if this is okay or whether we need
    // to pass in `config` and use `config.logger.log`.
    console.log(`${pjson.name} ${pjson.version}`);
    const versionInfo = await armlet.ApiVersion();
    console.log(versionJSON2String(versionInfo));

}

function setConfigSeverityLevel (inputSeverity) {

    // converting severity to a number makes it easier to deal with in `issues2eslint.js`
    const severity2Number = {
        'error': 2,
        'warning': 1
    };

    // default to `warning`
    return severity2Number[inputSeverity] || 1;
}

function setConfigSWCBlacklist (inputBlacklist) {
    if (!inputBlacklist) {
        return false;
    }

    return inputBlacklist
        .toString()
        .split(",")
        .map(swc => "SWC-" + swc.trim());
}

/**
 * Modifies attributes of the Truffle configuration object
 * in order to use project-defined options
 *
 * @param {Object} config - Truffle configuration object.
 * @returns {Oject} config - Extended Truffle configuration object.
 */

function prepareConfig (config) {

    // merge project level configuration
    let projectConfig;
    try {
        projectConfig = require([config.working_directory, 'truffle-security'].join ('/'));
    } catch (ex) {
        projectConfig = {}
        if (config.debug) {
            config.logger.log("truffle-security.json either not found or improperly formatted. Default options will be applied.");
        }
    }

    const projectLevelKeys = Object.keys(projectConfig);

    projectLevelKeys.forEach(function (property) {
        if (!config.hasOwnProperty(property)) {
            config[property] = projectConfig[property];
        }
    });

    // modify and extend initial config params
    config.severityThreshold = setConfigSeverityLevel(config['min-severity']);
    config.swcBlacklist = setConfigSWCBlacklist(config['swc-blacklist']);
    if (typeof(config['mythx-logs']) === 'undefined') {
      config.mythxLogs = true;
    }
    // API Client configuration
    if (typeof config.apiClient === undefined) {
        config.apiClient = defaultAPIClient;
    }

    return config;
}


module.exports = {
    analyze,
    defaultAnalyzeRateLimit,
    versionJSON2String,
    printVersion,
    printHelpMessage,
    setConfigSeverityLevel,
    setConfigSWCBlacklist,
    trialEthAddress,
    trialPassword,
};
